### PR TITLE
fix: [CI-7096]: Refactor TI client

### PR DESCRIPTION
**Pull request by rutvijmehta-harness**
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
    - CodeFormat: `trigger codeformat`
    - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- TI-ALL: `trigger tiAll`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/44112)

### CODE REVIEW

File: product/ci/engine/grpc/tiproxy_handler.go [hunk 1]
```diff
@@ -179,61 +125,28 @@ func (h *tiProxyHandler) WriteTests(stream pb.TiProxy_WriteTestsServer) error {
...
 	repo := req.GetRepo()
-	if repo == "" {
-		return res, fmt.Errorf("repo not present in request")
-	}
+	sha := req.GetSha()
+	commitLink := ""
+	skipVerify := false
+	tiClient := getRemoteTiClient(repo, sha, commitLink, skipVerify)
+
+	// SelectTests API call
```
Comment (hemanthmantri) [inline product/ci/engine/grpc/tiproxy_handler.go:135]: change?

File: product/ci/engine/grpc/tiproxy_handler.go [hunk 2]
```diff
@@ -315,31 +230,21 @@ func (h *tiProxyHandler) getEncodedData(req *pb.UploadCgRequest) ([]byte, string
...
-	if err != nil {
-		h.log.Errorw("could not create a client to the TI service", zap.Error(err))
-		return nil, err
-	}
+	// TI Client
+	repo := ""
+	sha := ""
+	commitLink := ""
+	skipVerify := false
+	tc := getRemoteTiClient(repo, sha, commitLink, skipVerify)
```
Comment (hemanthmantri) [inline product/ci/engine/grpc/tiproxy_handler.go:238]: tiClient - keep it consistent across all methods.

File: product/ci/ti-service/client/http.go [hunk 3]
```diff
@@ -131,77 +134,84 @@ type HTTPClient struct {
...
+	PipelineID string
+	BuildID    string
+	StageID    string
+	Repo       string
+	Sha        string
+	CommitLink string
 	SkipVerify bool
 }
 
 // Write writes test results to the TI server
```
Comment (hemanthmantri) [inline product/ci/ti-service/client/http.go:148]: Write test results

**Review by hemanthmantri — DISMISSED** [review]
The interface change means TI-service also has to understand the new inputs, right? How do we make sure accounts that are using old AddOn version are not broken? We should make TI server side changes backward compatible.

File: product/ci/ti-service/client/http.go (hunk 3 above)
Comment (rutvijmehta-harness) [reply product/ci/ti-service/client/http.go:148]: The comment format is <method name> <what it does>

File: product/ci/engine/grpc/tiproxy_handler.go (hunk 1 above)
Comment (rutvijmehta-harness) [reply product/ci/engine/grpc/tiproxy_handler.go]: Done

File: product/ci/engine/grpc/tiproxy_handler.go (hunk 2 above)
Comment (rutvijmehta-harness) [reply product/ci/engine/grpc/tiproxy_handler.go]: Done

**Review by rutvijmehta-harness — COMMENTED** [review]
> The interface change means TI-service also has to understand the new inputs, right? How do we make sure accounts that are using old AddOn version are not broken? We should make TI server side changes backward compatible.

There are no interface changes. The data transferred over network is still JSON which is read into a struct which understands this JSON. This is just a client refactor but I'll test out all the scenarios with old/new ti-service, addon and lite-engine when we import the new ti-client.

**Comment by rutvijmehta-harness** [comment]
Tests run:

New lite-engine / old addon / old ti service
full run - https://qa.harness.io/ng/#/account/uy6FLHeWTBOOPZA8xchgZA/ci/orgs/default/projects/Default_Project_1657212353481/pipelines/test/executions/dfzVQx23SdyIR1bGXulOCA/pipeline?storeType=INLINE
test selection - https://qa.harness.io/ng/#/account/uy6FLHeWTBOOPZA8xchgZA/ci/orgs/default/projects/Default_Project_1657212353481/pipelines/test/executions/Mkzmbgc6QVCRTalZ79zbDw/pipeline?storeType=INLINE

New lite-engine / new addon / old ti service
test selection - https://qa.harness.io/ng/#/account/uy6FLHeWTBOOPZA8xchgZA/ci/orgs/default/projects/Default_Project_1657212353481/pipelines/test/executions/NTTS88ISTT6TkiI-TbZzsA/pipeline?storeType=INLINE

**Comment by abhayganvir** [comment]
trigger  abhaysmartchecks

**Comment by abhayganvir** [comment]
trigger abhaysmartchecks

**Comment by abhayganvir** [comment]
trigger smartchecks

**Review by jamie-harness — APPROVED** [review]
> Reviewed 4 of 6 files at r1, 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44112)* status: 3. unresolved discussions.

**Review by hemanthmantri — APPROVED** [review]
:lgtm:

> Reviewed 4 of 6 files at r1, 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44112#-:-NOpVD5Q4WM4MvLmTX01:bnfp4nl)* status: :shipit: complete! ✓

**Pull request merged**